## Libraries

```{r libraries}
library(reshape)
library(plyr)
library(aroma.affymetrix)
```

## Variables

The aroma.affymetrix and this script library is very fussy about directory structure!

Affymetrix SNP 6 annotation file must be in the correct directory (annotationData) under the current working directory (getwd()). 
Like this:
    baseDir
      └── annotationData/
          └── chipTypes
              └── GenomeWideSNP_6.Full
                  └── GenomeWideSNP_6.Full.cdf
            
This script will search for raw data (CEL files) in the "CELfiles" subdirectory of the rawData directory under the current working directory. 

    baseDir
      └── rawData/
          └── CELfiles
              └── foo.CEL
              
Note that while it is possible to process all the CEL files at once, the process is very memory intensive, so this script processes the CEL files one at a time. 

```{r variables}
# the base of the aroma.affymetrix directory structure
baseDir <- "/Volumes/2TB/backups/Broad_backup/pipeline" # change this to match your own basedirectory
```

## Annotations
 - loads in probe annotations for the Affy 6 array

```{r annotations}
# load in probe annotations, detailing allele identities of the "A" and "B" probesetIDs
A.B.probe.annots <- read.table("./annotationData/simplified.annot.A.B.hg18.tab", header=T, sep="\t")
# reformat annotations and label
A.B.probe.annots.m <- melt(A.B.probe.annots, id.vars="SNP_ID")
names(A.B.probe.annots.m) <- c("probeset", "alleleID", "allele")

# load array library (locations of probes on chip etc.)
cdf <- AffymetrixCdfFile$byChipType("GenomeWideSNP_6.Full") 
# extract the indices of all SNP probes
SNPprobe.indices <- grep("SNP_A", getUnitNames(cdf))
```

## Process arrays and write to file

```{r process}
# list CEL files
celfiles <- list.files(file.path(baseDir, "rawData", "CELfiles"), pattern="CEL|cel" )

# load CEL files and process one at a time
sapply(celfiles, function(celfile) {
  cf <- AffymetrixCelFile(file.path(baseDir, "rawData", "CELfiles", celfile), cdf=cdf)
  cs <- AffymetrixCelSet(files=list(cf))
  # get individual probesetID intensities
  y <- getUnitIntensities(cs, units=SNPprobe.indices, verbose=T)
  # reformat list of lists into a informatively labelled dataframe 
intensities <- melt(y)
names(intensities)[c(1:2, 5,6)] <- c("probe", "sample", "allele", "probeset")
# take median value for all probes within a probesetID
intensities.median <- aggregate(intensities$value, by=list(sample=intensities$sample, allele=intensities$allele, probeset=intensities$probeset), median)
# merge in the AlleleIDs
intensities.median <- merge(intensities.median, A.B.probe.annots.m, by=c("allele", "probeset"))
intensities.median$probesetID <- paste(intensities.median$probeset, intensities.median$alleleID, sep="_") 

data <- cast(intensities.median, probesetID ~ sample, value="x")
```


Terminology (these are not the most consistent)
Probe - refers to an individual instance of a probe on the array for a particular SNP and allele, there are multiple probes with identical properties and targets placed across the array
Probeset - summarized group of probes against a particular SNP
i.e. SNP_A-1780415

ProbesetID - group of probes against a particular SNP AND allele
SNP_A-1780415_A
or 
SNP_A-1780415_B

alleleID - the A or B on the end of the probesetID